src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+      //you must call SS.open_* to replace de Bruijn indexes with names
+      let xs, body = SS.open_term xs body in
+      let patterns = xs |> List.map (fun (x, qual) ->
+        //x.sort contains a type annotation for the bound variable
+        //the pattern `p` below only contains the variable, not the annotation
+        //but, if the user wrote the annotation, then we should record that and print it back
+        //additionally, if we're in verbose mode, e.g., if --print_bound_var_types is set
+        //    then we should print the annotation too
+        let p = A.PatVar(bv_as_unique_ident x, resugar_arg_qual qual) in
+        A.mk_pattern p (S.range_of_bv x)) in
+      let body = resugar_term body in
+      mk (A.Abs(patterns, body))
+
+    | Tm_arrow(xs, body) ->
+      let xs, body = SS.open_comp xs body in
+      let body = resugar_comp body in
+      //let xs, body = SS.open_term xs body in
+      let xs = xs |> List.map (fun (b, qual) -> resugar_bv_as_binder b t.pos) in
+      let rec aux body = function
+        | [] -> body
+        | hd::tl ->
+          let body = mk (A.Product([hd], body)) in
+          aux body tl in
+      aux body xs
+
+    | Tm_refine(x, phi) ->
+      (* bv * term -> binder * term *)
+      let x, phi = SS.open_term [S.mk_binder x] phi in
+      let (b, _) = x.Head in 
+      let b = resugar_bv_as_binder b t.pos in
+      mk (A.Refine(b, resugar_term phi))
+
+    | Tm_app(e, args) ->
+      (* Op("=!=", args) is desugared into Op("~", Op("==") and not resugared back as "=!=" *)
+      begin match resugar_term_as_op e with
+        | None->
+          let args = args |> List.map (fun (e, qual) ->
+            resugar_term e) in
+          let e = resugar_term e in
+          let rec aux e = function
+            | [] -> e
+            | hd::tl ->
+              let e = mk (A.App(e, hd, A.Nothing)) in
+              aux e tl in
+          aux e args
+
+        | Some "tuple" ->
+          let args = args |> List.map (fun (e, qual) ->
+            resugar_term e) in
+          let rec aux e = function
+            | [] -> e
+            | hd::tl ->
+              let e = mk (A.Op("*", [e;hd])) in
+              aux e tl in
+          let e = mk(A.Op("*", [args.Head; args.Tail.Head])) in
+          aux e args.Tail.Tail
+
+        | Some "dtuple" -> 
+          (* this is desugared from Sum(binders*term) *)
+          let rec last = function
+            | hd :: [] -> hd
+            | hd :: tl -> last tl
+            | _ -> failwith "Empty list." in
+          let body, _ = last args in
+          let binders, body = match (SS.compress body).n with
+            | Tm_abs(xs, body, _) ->
+                let xs, body = SS.open_term xs body in
+                let xs = xs |> List.map (fun (b, qual) -> resugar_bv_as_binder b t.pos) in 
+                let body = resugar_term body in
+                xs, body
+            | _ ->
+              [], resugar_term body in
```

Comment:
We hit this case when we try to resugar ulib\prims.fst where it actually defines dtuple, in which case we will just resugar back as AST.App instead of AST.Sum.